stdlib/public/Concurrency/TaskGroup.cpp

```diff
@@ -641,37 +670,54 @@ void TaskGroupImpl::offer(AsyncTask *completedTask, AsyncContext *context) {
           waitingTask, nullptr,
           /*success*/ std::memory_order_release,
           /*failure*/ std::memory_order_acquire)) {
+
 #if SWIFT_CONCURRENCY_TASK_TO_THREAD_MODEL
-         // We have completed a child task in a task group task and we know
-         // there is a waiting task who will reevaluate TaskGroupImpl::poll once
-         // we return, by virtue of being in the task-to-thread model.
-         // We want poll() to then satisfy the condition of having readyTasks()
-         // that it can dequeue from the readyQueue so we need to enqueue our
-         // completion.
-
-         // TODO (rokhinip): There's probably a more efficient way to deal with
-         // this since the child task can directly offer the result to the
-         // parent who will run next but that requires a fair bit of plumbing
-         enqueueCompletedTask(completedTask, hadErrorResult);
-         unlock(); // TODO: remove fragment lock, and use status for synchronization
-         return;
+        // In the task-to-thread model, child tasks are always actually
+        // run synchronously on the parent task's thread.  For task groups
+        // specifically, this means that poll() will pick a child task
+        // that was added to the group and run it to completion as a
+        // subroutine.  Therefore, when we enter offer(), we know that
+        // the parent task is waiting and we can just return to it.
+
+        // The task-to-thread logic in poll() currently expects the child
+        // task to enqueue itself instead of just filling in the result in
+        // the waiting task.  This is a little wasteful; there's no reason
+        // we can't just have the parent task set itself up as a waiter.
+        // But since it's what we're doing, we basically take the same
+        // path as we would if there wasn't a waiter.
```

Comment:
Does this also want a Radar or something to track fixing it so the parent sets itself up as a waiter?

Reply:
This is a much better comment than what I wrote - thank you.